ethcore/src/executive.rs

```diff
@@ -276,7 +276,7 @@ impl<'a, B: 'a + StateBackend> Executive<'a, B> {
 
 			let cost = builtin.cost(data);
 			if cost <= params.gas {
-				builtin.execute(data, &mut output);
+				builtin.execute(data, &mut output)?;
```

Comment:
You are right, I misjudged the program flow here
the errors and rollback/discard logic should be decoupled from each other though